src/Microsoft.AspNet.Mvc.Core/Internal/ActionConstraintCache.cs

```diff
@@ -0,0 +1,213 @@
+                    items.Add(item);
+                }
+                else
+                {
+                    items.Add(new ActionConstraintItem(item.Metadata));
+                }
+            }
+
+            ExecuteProviders(httpContext, action, items);
+
+            return ExtractActionConstraints(items);
+        }
+
+        private void ExecuteProviders(HttpContext httpContext, ActionDescriptor action, List<ActionConstraintItem> items)
+        {
+            var context = new ActionConstraintProviderContext(httpContext, action, items);
+
+            for (var i = 0; i < _actionConstraintProviders.Length; i++)
+            {
+                _actionConstraintProviders[i].OnProvidersExecuting(context);
+            }
+
+            for (var i = _actionConstraintProviders.Length - 1; i >= 0; i--)
+            {
+                _actionConstraintProviders[i].OnProvidersExecuted(context);
+            }
+        }
+
+        private IReadOnlyList<IActionConstraint> ExtractActionConstraints(List<ActionConstraintItem> items)
+        {
+            var count = 0;
+            for (var i = 0; i < items.Count; i++)
+            {
+                if (items[i].Constraint != null)
+                {
+                    count++;
+                }
+            }
+
+            if (count == 0)
+            {
+                return null;
+            }
+            
+            var actionConstraints = new IActionConstraint[count];
+            for (int i = 0, j = 0; i < items.Count; i++)
+            {
+                var actionConstraint = items[i].Constraint;
+                if (actionConstraint != null)
+                {
+                    actionConstraints[j++] = actionConstraint;
+                }
+            }
+
+            return actionConstraints;
+        }
+
+        private class InnerCache
+        {
+            public InnerCache(int version)
+            {
+                Version = version;
+            }
+
+            public ConcurrentDictionary<ActionDescriptor, CacheEntry> Entries { get; } =
+                new ConcurrentDictionary<ActionDescriptor, CacheEntry>();
+
+            public int Version { get; }
+        }
+
+        private struct CacheEntry
+        {
+            public CacheEntry(IReadOnlyList<IActionConstraint> actionConstraints)
+            {
+                ActionConstraints = actionConstraints;
+                Items = null;
+            }
+
+            public CacheEntry(List<ActionConstraintItem> items)
+            {
+                Items = new ActionConstraintItem[items.Count];
+                for (var i = 0; i < Items.Length; i++)
+                {
+                    var item = items[i];
+                    if (item.IsReusable)
+                    {
+                        Items[i] = item;
+                    }
+                    else
+                    {
+                        Items[i] = new ActionConstraintItem(item.Metadata);
+                    }
+                }
```

Comment:
It looks like this is an extra defensive copy of `items` - the filter cache code may have had this too. Can you take a look and tell me if you agree?